src/main/java/pl/jagiellonian/Models/PolynomialAsMap.java

```diff
@@ -37,16 +48,198 @@ public PolynomialAsMap multipleWith(PolynomialAsMap myPolynomial) {
 
                 List<Integer> monomialPowersResult = new ArrayList<>();
 
-                for (int i = 0; i < thisMonomialPowers.size(); i++) {
-                    monomialPowersResult.add(thisMonomialPowers.get(i) + monomialPowers.get(i));
+                int i = 0;
+                for (; i < thisMonomialPowers.size(); i++) {
+                    if (monomialPowers.size() > i) {
+                        monomialPowersResult.add(thisMonomialPowers.get(i) + monomialPowers.get(i));
+                    } else {
+                        monomialPowersResult.add(thisMonomialPowers.get(i));
+                    }
+                }
+                for (; i < monomialPowers.size(); i++) {
+                    monomialPowersResult.add(monomialPowers.get(i));
                 }
 
                 Integer monomialResultCoefficient = thisMonomialCoefficient * monomialCoefficient;
                 polynomialResult.put(monomialPowersResult, monomialResultCoefficient);
             }
         }
-        PolynomialAsMap polynomialAsMap = new PolynomialAsMap(polynomialResult);
-        return polynomialAsMap;
+        return new PolynomialAsMap(polynomialResult);
+    }
+
+    public PolynomialAsMap addPolynomial(PolynomialAsMap polynomialAsMap) {
+        Map<List<Integer>, Integer> polynomialToAddMap = new HashMap<>(polynomialAsMap.getPolynomialMap());
+        for (List<Integer> integers : polynomialMap.keySet()) {
+            if (polynomialToAddMap.get(integers) != null) {
+                polynomialToAddMap.put(integers, polynomialToAddMap.get(integers) + polynomialMap.get(integers));
```

Comment:
addPolynomial makes inefficient use of keySet iterator instead of entrySet iterator